http-client/Network/HTTP/Client/Body.hs

```diff
@@ -115,9 +115,13 @@ makeUnlimitedReader :: Connection -> IO BodyReader
 makeUnlimitedReader Connection {..} = do
     icomplete <- newIORef False
     return $ do
-        bs <- connectionRead
-        when (S.null bs) $ writeIORef icomplete True
-        return bs
+      complete <- readIORef icomplete
+      if complete
+          then return empty
+          else do
+              bs <- connectionRead
+              when (S.null bs) $ writeIORef icomplete True
+              return bs
```

Comment:
By now we have:

https://github.com/snoyberg/http-client/blob/ed974a7af8adeb75f4c02b80d3f2bec4857b9f37/http-client/Network/HTTP/Client/Body.hs#L108-L119

Given the `cleanup` action , I think this is even more crucial now.